### PR TITLE
tikv_kv: make async_snapshot return future

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?
Issue Number: Ref #13827.

What's Changed:
```commit-message
so we can reduce allocation and keep compatible with both v1 and v2.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by BusyJay** [comment]
Microbench shows that when both using `tikv_kv::snapshot`, the worst case (always return error) time cost is reduced from 960ns to 920ns. But if using `raftkv::async_snapshot` directly, the worst case time cost is increased from 920ns to 818ns. The difference due to the new implementation always use `MustCall` as its only callback, which is more costy than naive callback.

File: components/tikv_kv/src/lib.rs [hunk 1]
```diff
@@ -302,17 +304,21 @@ pub trait Engine: Send + Clone + 'static {
...
-        wait_op!(|cb| self.async_snapshot(ctx, cb), timeout)
-            .unwrap_or_else(|| Err(Error::from(ErrorInner::Timeout(timeout))))
+        let deadline = Instant::now() + DEFAULT_TIMEOUT;
+        let timeout = GLOBAL_TIMER_HANDLE.delay(deadline).compat();
+        futures::executor::block_on(async move {
+            futures::select! {
+                res = self.async_snapshot(ctx).fuse() => res,
+                _ = timeout.fuse() => Err(Error::from(ErrorInner::Timeout(DEFAULT_TIMEOUT))),
+            }
+        })
```
Comment (sticnarf) [inline components/tikv_kv/src/lib.rs:321]: Can we use `GLOBAL_TIMER_HANDLE.timeout()`? That looks simpler.

Comment (BusyJay) [reply components/tikv_kv/src/lib.rs:321]: `timeout` is v1, so the future will be converted twice.

Comment (sticnarf) [reply components/tikv_kv/src/lib.rs:321]: After creating the timeout, we can just call `wait` on the v1 future.

File: src/server/raftkv.rs [hunk 2]
```diff
@@ -481,10 +458,46 @@ where
         }
         ASYNC_REQUESTS_COUNTER_VEC.snapshot.all.inc();
         let begin_instant = Instant::now_coarse();
-        self.exec_snapshot(
-            ctx,
-            req,
-            Box::new(move |res| match res {
+        let (cb, f) = paired_must_called_future_callback(drop_snapshot_callback);
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/server/raftkv.rs:461]: I prefer to keep the v1 code as it is.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: components/tikv_kv/src/lib.rs (hunk 1 above)
Comment (BusyJay) [reply components/tikv_kv/src/lib.rs:321]: Then it will require `SnapshotRes` to be `unpin`.

**Review by sticnarf — APPROVED** [review]

**Pull request merged**